mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -0,0 +1,436 @@
+        input_channels (int, optional): Channels of the input image.
+            Defaults to 3.
+        downsample_cfg (list[bool] | dict[list], optional): Config for
+            downsample behavior of the intermedia layers. If a list is passed,
+            ``downsample_cfg[idx] == True`` means apply downsample in idx-th
+            block, and vice versa. If dict is provided, the key dict should
+            be the input scale of the image and corresponding value should be
+            a list ti define the downsample behavior. Default: Please refer
+            to ``_default_downsample_cfg``.
+        from_rgb_cfg (dict, optional): Config for the first layer to convert
+            rgb image to feature map. Defaults to
+            ``dict(type='SNGANDiscHeadResBlock')``.
+        blocks_cfg (dict, optional): Config for the intermedia blocks.
+            Defaults to ``dict(type='SNGANDiscResBlock')``
+        act_cfg (dict, optional): Activation config for the final output
+            layer. Defaults to ``dict(type='ReLU')``.
+        with_spectral_norm (bool, optional): Whether use spectral norm for
+            all conv blocks or not. Default to True.
+    """
+
+    _defualt_channels_cfg = {
+        32: [1, 1, 1],
+        64: [2, 4, 8, 16],
+        256: [2, 4, 8, 8, 16, 16]
+    }
+
+    _defualt_downsample_cfg = {
+        32: [True, False, False],
+        64: [True, True, True, True],
+        256: [True, True, True, True, True, False]
+    }
+
+    def __init__(self,
+                 input_scale,
+                 base_channels,
+                 num_classes=0,
+                 input_channels=3,
+                 channel_cfg=None,
+                 downsample_cfg=None,
+                 from_rgb_cfg=dict(type='SNGANDiscHeadResBlock'),
+                 blocks_cfg=dict(type='SNGANDiscResBlock'),
+                 act_cfg=dict(type='ReLU'),
+                 with_spectral_norm=True):
+        """"""
+        super().__init__()
+
+        # add SN options and activation function options to cfg
+        self.from_rgb_cfg = deepcopy(from_rgb_cfg)
+        self.from_rgb_cfg.setdefault('act_cfg', act_cfg)
+        self.from_rgb_cfg.setdefault('with_spectral_norm', with_spectral_norm)
+
+        # add SN options and activation function options to cfg
+        self.blocks_cfg = deepcopy(blocks_cfg)
+        self.blocks_cfg.setdefault('act_cfg', act_cfg)
+        self.blocks_cfg.setdefault('with_spectral_norm', with_spectral_norm)
+
+        channel_cfg = deepcopy(self._defualt_channels_cfg) \
+            if channel_cfg is None else deepcopy(channel_cfg)
+        if isinstance(channel_cfg, dict):
+            if input_scale not in channel_cfg:
+                raise KeyError(f'`input_scale={input_scale} is not found in '
+                               '`channel_cfg`, only support configs for '
+                               f'{[chn for chn in channel_cfg.keys()]}')
+            self.channel_factor_list = channel_cfg[input_scale]
```

Comment:
Once `not isinstance(channel_cfg, dict)`, it seems that `channel_factor_list` will not exist.